XMonad/Hooks/BorderPerWindow.hs

```diff
@@ -0,0 +1,132 @@
+-- >     , ...
+-- >     }
+
+-- $design
+--
+-- 1. Keep it simple. Since the extension does not aim to change border setting
+--    when layout changes, only execute the border setting function once to
+--    avoid potential window flashing/jumping/scaling.
+--
+-- 2. The 'ManageHook' eDSL is a nice language for specifying windows. Let's
+--    build on top of it and use it to specify window to define border.
+
+-- $options
+--
+-- 1. Just setting window-specific border in 'manageHook' does not work, since
+--    XMonad sets window border after 'ManageHook's are applied.
+--
+-- 2. 'XMonad.Layout.NoBorders' provides a way to set zero-width window border
+--    not permanently, e.g., need to recover border when layout changes, which
+--    results in complexity, e.g., need to set border every time a layout is
+--    rendered, and has a window jumping effect when switching between
+--    workspaces: for a zero-width border window, the border (global config) is
+--    temporarily drawn and then removed; all pixels of the window is scaled
+--    during the add-n-remove procedure, which could be noticeable.
+--
+-- 3. 'XMonad.Layout.VoidBorders' allows one to choose layout but not window.
+
+-- | The core type that records border width per window
+data WindowBorderSpec
+    = WindowBorderSpec Window Dimension deriving (Read, Show)
+
+-- | The XMonad message that transfers 'WindowBorderSpec' from the window
+-- creation event to the layout handling function.
+-- Note that XMonad reads config and draws window border after 'ManageHook's
+-- are applied. It does not work if we set the border in a 'ManageHook'
+-- (hooking window creation). So we choose to send a message in
+-- 'defineWindowBorder' function to a later stage, e.g., layout handling, to
+-- set the window border.
+data WindowBorderSpecMessage
+    = WindowBorderSpecMessage WindowBorderSpec deriving (Read, Show)
+
+-- | Define 'WindowBorderSpecMessage' as a XMonad 'Message'
+instance Message WindowBorderSpecMessage
+
+-- | The function to specify border width in a 'ManageHook'.
+defineWindowBorder :: Dimension -> ManageHook
+defineWindowBorder bw = do
+    w <- ask
+    liftX . broadcastMessage . WindowBorderSpecMessage $ WindowBorderSpec w bw
+    idHook
+
+-- | The core XMonad 'LayoutModifier' type for this extension. This will be
+-- used by our layout modifier function 'setBorderPerWindow' to execute the
+-- procedure of setting border for each window according to the
+-- 'WindowBorderSpecMessage' message received for that window.
+data BorderPerWindow a
+    = BorderPerWindow (Maybe WindowBorderSpec) deriving (Read, Show)
+
+instance LayoutModifier BorderPerWindow a where
+    modifierDescription = const "BorderPerWindow"
+
+    redoLayout (BorderPerWindow Nothing) _ _ wrs =
+        return (wrs, Nothing)
+
+    redoLayout (BorderPerWindow (Just (WindowBorderSpec w bw))) _ _ wrs = do
+        withDisplay $ \d -> io $ setWindowBorderWidth d w bw
+        return (wrs, Nothing)
```

Comment:
I think this would be an ideal candidate for separating the `X` part and the layouting part.  Basically, instead of specifying `redoLayout` you could put the `withDisplay` call into `hook` and not have all of this noise about returning the window/rectangle pairs (this is then taken care of by the default implementation of `pureModifier` and `redoLayout`)

Reply:
I need to make sure I fully understand your point. Please correct me if I misunderstand any part of it.

If we put `withDisplay` (basically `setWindowBorderWidth`) into `hook` (I assume you mean the manage hook `defineWindowBorder`), then we actually do not need a layout hook afterwards, so it goes to [Attempt A in my stackoverflow question](https://stackoverflow.com/questions/69725874/config-no-border-for-specific-windows-in-xmonad) but [people found it does not work 12 years ago](https://xmonad.haskell.narkive.com/130DKJDZ/how-to-make-a-window-floating-above-all-the-other-windows-and-border-less) and I checked `XMonad.operations` code and [confirmed it does not work](https://stackoverflow.com/a/69774944/8418198). As I detailed in [my comment below](https://github.com/xmonad/xmonad-contrib/pull/640#issuecomment-957847814), basically I need to execute `setWindowBorderWidth` outside and after a manage hook to make it work, so I create a layout hook to do it and try to carefully only execute it once (not to have any [windows flashing/jumping/scaling effect](https://github.com/xmonad/xmonad-contrib/issues/316)).

Please advice or correct me if I incorrectly understand your point.

Reply:
Sorry, I should've been more clear.  The `LayoutModifier` class actually has a [hook](https://hackage.haskell.org/package/xmonad-contrib-0.17.0/docs/XMonad-Layout-LayoutModifier.html#v:hook) member function (perhaps a bit weirdly named) which gets executed by the default implementation of `redoLayout`: 

``` haskell
  redoLayout m r ms wrs = do hook m; return $ pureModifier m r ms wrs
```

Since we're not doing any layouty bits at all (i.e., we just return `(wrs, Nothing)`) and the only interesting thing is the call to `setWindowBorderWidth`, I was suggesting to implement `hook` instead of `redoLayout` here

Reply:
I see. Thank you for the information! Yes, this is a better place. Let me try this as well as `logHook` to figure out a good way to implement the functionality.